### PR TITLE
Base tests for parser correctness

### PR DESCRIPTION
**Pull request by feds01**
This PR adds a set of basic tests to verify that the parser is able to correctly handle various edge cases, states and forms. This is likely not a complete set of feature tests and more will be added in the future. However, this set of tests is good enough to now be merged into mainstream, and be used by commit to verify the correctness of the parser and eliminate regression. Additionally, this PR drops support for the PEST parser as it's considered that the self-hosted parser is now fully-fledged, more accurate, faster and has a considerably lower memory footprint.

During the test writing, various bugs were discovered and fixed which are detailed below:
- better error messages for unclosed token trees.
- fix panic when a particular token tree was unclosed and at the end of the file.
- fix: don't put keywords in the identifier map as it increases the amount of work done by the hashmap lookups (saves about ~10ms avg).
- fix panic that occurs when reporting crate tries to print an error for module contents that don't exist. This mainly occurred to an edge case of an early failing condition and using default `Location` when converting an `ImportError` into a `ParseError`.
- fix very confusing error messages occurring inside of token trees that were being printed from the parent's context whereas they should have been referencing the child context.
- fix set and map literal parsing confusion with block statements.
- fix weirdness with struct literals and for-loops
- fix weirdness with re-assignment operators not being allowed in some cases for blocks.
- add support for type arguments when a function call is written in infix style.
- fix unclosed string literals not being shown as errors in the tokeniser.
- fix float literals not being able to be parsed with '_' in the decimal part.
- fix enforce Unicode escape literals have a maximum of 6 hex chars.
- fix support re-assignment in parentheses and other locations.
- fix dangling token trees allowed if the previous token tree was of the same delimiter type.

Miscellaneous improvements:
- Use Rust 2021
- Upgrade all project dependencies
- Added support struct literal field name shortcut de-sugaring (ecd5527aa8a2d9344e5bba6f9fcc1e393ddf2509)
- standardise the logging label style so that it conforms to the rest of the project style.
- Add `SliceDisplay` in `hash-utils` (moved from the `hash-parser` crate) for pretty-printing vectors in an English-like form.

### CODE REVIEW

File: compiler/hash-ast/src/resolve.rs [hunk 1]
```diff
@@ -129,20 +129,20 @@ where
...
+                //         contents of the file for reporting purposes.
+                ctx.module_builder
+                    .add_contents(import_index, resolved_import_path, import_source);
+
+                match import_node {
+                    Ok(module) => {
+                        ctx.module_builder.add_module_at(import_index, module); // Add the import to modules
+                        Ok(())
+                    }
+                    Err(err) => Err(err),
```
Comment (kontheocharis) [inline compiler/hash-ast/src/resolve.rs:144]: ? syntax could be used here.

File: compiler/hash-utils/src/printing.rs [hunk 2]
```diff
@@ -0,0 +1,39 @@
+use std::fmt;
+
+/// Type wrapper around a slice of items.
+pub struct SliceDisplay<'a, T: 'a>(pub &'a [T]);
```
Comment (kontheocharis) [inline compiler/hash-utils/src/printing.rs:4]: Given the docs below, maybe it would be better to call this "SequenceDisplay" or something, not the programming-specific "SliceDisplay". Also since this is the purpose of this struct, maybe the doc comment should be on the struct itself and not on the fmt::Display implementation.

File: tests/parser/benches/benches.rs [hunk 3]
```diff
@@ -0,0 +1,40 @@
+#![feature(test)]
+extern crate test;
+
+use hash_alloc::Castle;
+use hash_ast::module::ModuleIdx;
+use hash_parser::lexer::Lexer;
+use test::{black_box, Bencher};
+
+static STRINGS: &str = r#""tree" "to" "a" "graph" "that can" "more adequately represent" "loops and arbitrary state jumps" "with\"\"\"out" "the\n\n\n\n\n" "expl\"\"\"osive" "nature\"""of trying to build up all possible permutations in a tree." "tree" "to" "a" "graph" "that can" "more adequately represent" "loops and arbitrary state jumps" "with\"\"\"out" "the\n\n\n\n\n" "expl\"\"\"osive" "nature\"""of trying to build up all possible permutations in a tree." "tree" "to" "a" "graph" "that can" "more adequately represent" "loops and arbitrary state jumps" "with\"\"\"out" "the\n\n\n\n\n" "expl\"\"\"osive" "nature\"""of trying to build up all possible permutations in a tree." "tree" "to" "a" "graph" "that can" "more adequately represent" "loops and arbitrary state jumps" "with\"\"\"out" "the\n\n\n\n\n" "expl\"\"\"osive" "nature\"""of trying to build up all possible permutations in a tree.""#;
```
Comment (kontheocharis) [inline tests/parser/benches/benches.rs:9]: Wow

File: tests/parser/tests/cases/should_fail/destructure_onto_void/case.hash [hunk 4]
```diff
@@ -0,0 +1,2 @@
+// This doesn't make much sense semantically either.
+let (,) = k;
```
Comment (kontheocharis) [inline tests/parser/tests/cases/should_fail/destructure_onto_void/case.hash:2]: Why should this fail? Can we not match on empty tuples?

File: tests/parser/tests/cases/should_fail/integer_literal_too_large/case.hash [hunk 5]
```diff
@@ -0,0 +1 @@
+let k = 23384912398412397432849123413027948091234809213480923409823231498;
```
Comment (kontheocharis) [inline tests/parser/tests/cases/should_fail/integer_literal_too_large/case.hash:1]: What is the maximum integer literal supported?

Comment (kontheocharis) [reply tests/parser/tests/cases/should_fail/integer_literal_too_large/case.hash:1]: Also, what about negative literals?

File: tests/parser/tests/cases/should_fail/non_hex_unicode_char/case.hash [hunk 6]
```diff
@@ -0,0 +1 @@
+let k = '\u{g00000}';
```
Comment (kontheocharis) [inline tests/parser/tests/cases/should_fail/non_hex_unicode_char/case.hash:1]: Everybody gangsta until g shows up in hex

File: tests/parser/tests/cases/should_pass/compount_pattern_matching_with_structs/case.hash [hunk 7]
```diff
@@ -0,0 +1,6 @@
+let data = match some_struct {
+    Dog { name, age = 6 } | Dog { name, age = 5 }  => do_something_with_name();
+    Dog { name, age = 13 } | Dog { name, age } if age % 2 == 0  => do_something_with_name();
+    Dog { name, age } => do_something_with_name();
+    _ => #unreachble();
+};
```
Comment (kontheocharis) [inline tests/parser/tests/cases/should_pass/compount_pattern_matching_with_structs/case.hash:6]: This is neat! (though you misspelled unreachable)

File: tests/parser/tests/cases/should_pass/if_branches/case.hash [hunk 8]
```diff
@@ -0,0 +1,7 @@
+if b == 2 {
+     print("b is 2")
+} else if b == 3 {
+     print("b is 3")
+} else {
+    print("b isn't 2 or 3 ")
+}
```
Comment (kontheocharis) [inline tests/parser/tests/cases/should_pass/if_branches/case.hash:7]: Is this meant to be an expression? How does it pass without semis?

File: tests/parser/tests/cases/should_pass/inline_comment/case.hash [hunk 9]
```diff
@@ -0,0 +1,8 @@
+if b == 2 {
+     print("b is 2")
+} else if b == 3 {
+     // print("b is 3")
+     a
+} else {
+    print("b isn't 2 or 3 ")
+}
```
Comment (kontheocharis) [inline tests/parser/tests/cases/should_pass/inline_comment/case.hash:8]: See previous.

File: tests/parser/tests/cases/should_pass/nested_re_assignment/case.hash [hunk 10]
```diff
@@ -0,0 +1,2 @@
+// This technically should parse, but it wouldn't make sense since the '+=' operator has a void return type
```
Comment (kontheocharis) [inline tests/parser/tests/cases/should_pass/nested_re_assignment/case.hash:1]: Is that a part of the trait definition?

File: tests/parser/tests/cases/should_pass/parenthesised_for_loop/case.hash [hunk 11]
```diff
@@ -0,0 +1,3 @@
+(for name in names.iter() {
```
Comment (kontheocharis) [inline tests/parser/tests/cases/should_pass/parenthesised_for_loop/case.hash:1]: Why should this pass?

File: tests/parser/tests/cases/should_pass/parenthesised_while_loop/case.hash [hunk 12]
```diff
@@ -0,0 +1,6 @@
+(
```
Comment (kontheocharis) [inline tests/parser/tests/cases/should_pass/parenthesised_while_loop/case.hash:1]: See for loop comment.

File: tests/parser/tests/cases/should_pass/pattern_matching_let/case.hash [hunk 13]
```diff
@@ -0,0 +1,6 @@
+// struct desturcturing
+let Dog { name } = dog;
+let Dog { name = better_name } = dog;
+
+// namespace destructuring
+let { name: even_better_name} = dog;
```
Comment (kontheocharis) [inline tests/parser/tests/cases/should_pass/pattern_matching_let/case.hash:6]: What do you think about having "=" for namespace destructuring, same as structs?

File: tests/parser/tests/cases/should_pass/refutable_pattern_destructuring/case.hash [hunk 14]
```diff
@@ -0,0 +1,6 @@
+// Well technically this pattern is 'irrefutable', even if we know that this is true, the compiler
+// cannot sensibly reason about it much further than parsing...
```
Comment (kontheocharis) [inline tests/parser/tests/cases/should_pass/refutable_pattern_destructuring/case.hash:2]: Unless we have literal types :)

File: tests/parser/tests/cases/should_pass/struct_pattern_matching/case.hash [hunk 15]
```diff
@@ -0,0 +1,5 @@
+let data = match some_struct {
+    Dog { name, age = 6 } => do_something_with_name();
+    Dog { name, age } => do_something_with_name();
+    _ => #unreachble();
```
Comment (kontheocharis) [inline tests/parser/tests/cases/should_pass/struct_pattern_matching/case.hash:4]: Same typo

File: tests/parser/tests/cases/should_pass/unary_ops/case.hash [hunk 16]
```diff
@@ -0,0 +1,3 @@
+let k = ~a;
+let k = ---a;
```
Comment (kontheocharis) [inline tests/parser/tests/cases/should_pass/unary_ops/case.hash:2]: In the future, we probably want a lint to say "Hey, --a does not mean to decrement a!".

**Review by kontheocharis — APPROVED** [review]
incroyable!

File: tests/parser/tests/cases/should_fail/destructure_onto_void/case.hash (hunk 4 above)
Comment (feds01) [reply tests/parser/tests/cases/should_fail/destructure_onto_void/case.hash:2]: Yes, this can be added as support in the parser although it doesn't make much sense to do this.

File: tests/parser/tests/cases/should_fail/integer_literal_too_large/case.hash (hunk 5 above)
Comment (feds01) [reply tests/parser/tests/cases/should_fail/integer_literal_too_large/case.hash:1]: The maximum integer supported is a `u64`, however, it maybe problematic with signed literals as then it becomes 2^6 3

File: tests/parser/tests/cases/should_pass/refutable_pattern_destructuring/case.hash (hunk 14 above)
Comment (feds01) [reply tests/parser/tests/cases/should_pass/refutable_pattern_destructuring/case.hash:2]: yeees

File: tests/parser/tests/cases/should_pass/unary_ops/case.hash (hunk 16 above)
Comment (feds01) [reply tests/parser/tests/cases/should_pass/unary_ops/case.hash:2]: that's a good point, it should be explicitly stated in the docs that it doesn't do what you think it does.

File: tests/parser/tests/cases/should_pass/if_branches/case.hash (hunk 8 above)
Comment (feds01) [reply tests/parser/tests/cases/should_pass/if_branches/case.hash:7]: well because it is treated as a block and thus it becomes an expression in the block. It might also be the case that it is at the end of the file.

File: tests/parser/tests/cases/should_pass/nested_re_assignment/case.hash (hunk 10 above)
Comment (feds01) [reply tests/parser/tests/cases/should_pass/nested_re_assignment/case.hash:1]: No i don't think it should be, ill remove this comment.

File: tests/parser/tests/cases/should_fail/integer_literal_too_large/case.hash (hunk 5 above)
Comment (feds01) [reply tests/parser/tests/cases/should_fail/integer_literal_too_large/case.hash:1]: We could get around this by not converting to negative straight away and using the `neg` trait?

File: compiler/hash-ast/src/resolve.rs (hunk 1 above)
Comment (feds01) [reply compiler/hash-ast/src/resolve.rs:144]: Well unfortunately the call still has to be made to add the contents of the file into the module builder:
```rs
 ctx.module_builder.add_contents(import_index, resolved_import_path, import_source);
```

If the call is made before, the string is moved and the borrow checking complains about this.

File: tests/parser/tests/cases/should_pass/parenthesised_for_loop/case.hash (hunk 11 above)
Comment (feds01) [reply tests/parser/tests/cases/should_pass/parenthesised_for_loop/case.hash:1]: Treated as an expression at the end of the module.

File: tests/parser/tests/cases/should_pass/pattern_matching_let/case.hash (hunk 13 above)
Comment (feds01) [reply tests/parser/tests/cases/should_pass/pattern_matching_let/case.hash:6]: I think it would be a lot better to standardise syntax and make the parsing code a little cleaner.

File: compiler/hash-ast/src/resolve.rs (hunk 1 above)
Comment (kontheocharis) [reply compiler/hash-ast/src/resolve.rs:144]:
```rs
import_node.map(|module| ctx.module_builder.add_module_at(import_index, module))
```
Doesn't this do the same thing?

File: tests/parser/tests/cases/should_fail/destructure_onto_void/case.hash (hunk 4 above)
Comment (kontheocharis) [reply tests/parser/tests/cases/should_fail/destructure_onto_void/case.hash:2]: Well, you're right it doesn't make sense now, but it might make sense if we ever have type unions. I guess for now it can be left like this.

File: tests/parser/tests/cases/should_fail/integer_literal_too_large/case.hash (hunk 5 above)
Comment (kontheocharis) [reply tests/parser/tests/cases/should_fail/integer_literal_too_large/case.hash:1]: hmm, true. Maybe that is indeed a good idea. However, negative integer literals won't be able to be represented up to -(2^64-1) anyway in the language (it will be -(2^63-1) again).

File: tests/parser/tests/cases/should_pass/if_branches/case.hash (hunk 8 above)
Comment (kontheocharis) [reply tests/parser/tests/cases/should_pass/if_branches/case.hash:7]: But doesn't an expression in statement position (since this is a module [right?]) require a semicolon at the end?

File: tests/parser/tests/cases/should_pass/parenthesised_for_loop/case.hash (hunk 11 above)
Comment (kontheocharis) [reply tests/parser/tests/cases/should_pass/parenthesised_for_loop/case.hash:1]: Okay, resolving this, see previous response.

File: tests/parser/tests/cases/should_pass/pattern_matching_let/case.hash (hunk 13 above)
Comment (kontheocharis) [reply tests/parser/tests/cases/should_pass/pattern_matching_let/case.hash:6]: Yeah I agree.

File: tests/parser/tests/cases/should_pass/unary_ops/case.hash (hunk 16 above)
Comment (kontheocharis) [reply tests/parser/tests/cases/should_pass/unary_ops/case.hash:2]: Maybe an issue can be created for that.

File: tests/parser/tests/cases/should_pass/pattern_matching_let/case.hash (hunk 13 above)
Comment (feds01) [reply tests/parser/tests/cases/should_pass/pattern_matching_let/case.hash:6]: Ill make an issue after this is merged.

File: tests/parser/tests/cases/should_pass/if_branches/case.hash (hunk 8 above)
Comment (feds01) [reply tests/parser/tests/cases/should_pass/if_branches/case.hash:7]: Not if it's a block

File: compiler/hash-ast/src/resolve.rs (hunk 1 above)
Comment (feds01) [reply compiler/hash-ast/src/resolve.rs:144]: `add_module_at` doesn't add the contents of the source for the very reason.
